src/xwayland/xwm/mod.rs

```diff
@@ -2118,22 +2098,50 @@ fn handle_event<D: XwmHandler + 'static>(
                         .iter_mut()
                         .find(|x| x.window_id() == msg.window || x.mapped_window_id() == Some(msg.window))
                     {
-                        // We get a WL_SURFACE_ID message when Xwayland creates a WlSurface for a
-                        // window. Both the creation of the surface and this client message happen at
-                        // roughly the same time and are sent over different sockets (X11 socket and
-                        // wayland socket). Thus, we could receive these two in any order. Hence, it
-                        // can happen that we get None below when X11 was faster than Wayland.
-
-                        let wl_surface = xwm.wl_client.object_from_protocol_id::<WlSurface>(&xwm.dh, wid);
-                        match wl_surface {
-                            Err(_) => {
-                                xwm.unpaired_surfaces.insert(wid, msg.window);
-                            }
-                            Ok(wl_surface) => {
-                                let surface = surface.clone();
-                                drop(_guard);
-                                X11Wm::new_surface(state, xwm_id, surface, wl_surface);
-                            }
+                        // This is the old, deprecated method for associating a
+                        // wl_surface with an X11 window. It can race with the
+                        // lifecycle of the wl_surface. Still, we can make it
+                        // available to compositors who want to support old
+                        // versions of xwayland by manually matching on ID in an
+                        // idle callback.
+                        surface.state.lock().unwrap().wl_surface_id = Some(wid);
+                    }
+                }
+                x if x == xwm.atoms.WL_SURFACE_SERIAL => {
+                    // This handles the case that the serial was already set on
+                    // the wayland side. For the reverse order, we have a commit
+                    // hook.
+                    if let Some(xsurface) = xwm
+                        .windows
+                        .iter_mut()
+                        .find(|x| x.window_id() == msg.window || x.mapped_window_id() == Some(msg.window))
+                    {
+                        drop(_guard);
+
+                        let serial_lo = msg.data.as_data32()[0];
+                        let serial_hi = msg.data.as_data32()[1];
+                        let serial = u64::from(serial_lo) | (u64::from(serial_hi) << 32);
+
+                        let surface_state = xsurface.state.clone();
+                        let mut guard = surface_state.lock().unwrap();
+                        guard.wl_surface_serial = Some(serial);
+
+                        xwm.unpaired_surfaces.insert(serial, msg.window);
```

Comment:
This is really unfortunate. We should only be setting this in the `else` below, but that would mean double-borrowing `state`. I stared at this for tens of minutes without seeing a way around it.

Reply:
I guess you could re-borrow the `xwm` after getting the surface, but meh.. this is fine.